### PR TITLE
Fix route wrap

### PR DESCRIPTION
**Pull request by wxiaoguang**
Because we introduced the new code:

* https://github.com/wxiaoguang/gitea/pull/3

We must make sure every `converted handler` can return `done=true` correctly.

Otherwise, there will be duplicated calls to `others` again and again. Known problems caused by this bug:
* visit `/api/swagger` triggers panic
* many APIs return duplicated response

```
func wrapInternal(handlers []wrappedHandlerFunc) http.HandlerFunc {
	return func(resp http.ResponseWriter, req *http.Request) {
		var defers []func()
		defer func() {
			for i := len(defers) - 1; i >= 0; i-- {
				defers[i]()
			}
		}()
		for i := 0; i < len(handlers); i++ {
			handler := handlers[i]
			others := handlers[i+1:]
			done, deferrable := handler(resp, req, others...)
			if deferrable != nil {
				defers = append(defers, deferrable)
			}
			if done {
				return
			}
		}
	}
}
```

### CODE REVIEW

File: modules/util/reflect.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+	"runtime"
+)
+
+// GetFunctionInfo is for debug purpose (ex: show a router handler's real name)
+func GetFunctionInfo(i interface{}) string {
+	fp := reflect.ValueOf(i).Pointer()
+	fi := runtime.FuncForPC(fp)
+	file, line := fi.FileLine(fp)
+	return fmt.Sprintf("%s(%s:%d)", fi.Name(), file, line)
+}
```
Comment (zeripath) [inline modules/util/reflect.go]: this seems unrelated?

**Review by zeripath — APPROVED** [review]

File: modules/util/reflect.go (hunk 1 above)
Comment (wxiaoguang) [reply modules/util/reflect.go]: For further debug, I believe we need it soon ...

**Review by Gusted — APPROVED** [review]
